### ISSUE TITLE
Installation of Spark 2.3.0 fails

### ISSUE THREAD

**Issue by meatballs** [pixiedust/pixiedust#714]
As a user, I want to install pixiedust so that I can use it.

### Expected behavior

Following the installation instructions leads to a successful installation

### Actual behavior

Running `jupyter pixiedust install` and selecting the default Spark version (2.3.0) gives a 404 error

### Steps to reproduce the behavior

Follow the installation instructions

**Comment by jerryasher** [comment]
I am receiving this error as well. 

I am using CPython native for Windows 3.7 on Win 10 

```
jerry@Dell5570 MINGW64 ~
$ pip list | grep jupy
jupyter            1.0.0
jupyter-client     5.2.3
jupyter-console    5.2.0
jupyter-core       4.4.0

jerry@Dell5570 MINGW64 ~
$ python --version
Python 3.7.0

jerry@Dell5570 MINGW64 ~
$ pip list | grep req
requests           2.19.1

jerry@Dell5570 MINGW64 ~
$
```

Specifically, the stacktrace looks like:

```
$ jupyter pixiedust install
Step 1: PIXIEDUST_HOME: C:\Users\jerry\pixiedust
        Keep y/n [y]? y
Step 2: SPARK_HOME: C:\Users\jerry\pixiedust\bin\spark
        Keep y/n [y]? y
Directory C:\Users\jerry\pixiedust\bin\spark does not contain a valid SPARK install
        Download Spark y/n [y]? y
What version would you like to download? 1.6.3, 2.0.2, 2.1.0, 2.2.0, 2.3.0 [2.3.0]:
Traceback (most recent call last):
  File "c:\python\python37\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\python\python37\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Python\Python37\Scripts\jupyter-pixiedust.EXE\__main__.py", line 9, in <module>
  File "c:\python\python37\lib\site-packages\install\pixiedustapp.py", line 41, in main
    PixiedustJupyterApp.launch_instance()
  File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 657, in launch_instance
    app.initialize(argv)
  File "<decorator-gen-2>", line 2, in initialize
  File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 296, in initialize
    self.parse_command_line(argv)
  File "<decorator-gen-4>", line 2, in parse_command_line
  File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 514, in parse_command_line
    return self.initialize_subcommand(subc, subargv)
  File "<decorator-gen-3>", line 2, in initialize_subcommand
  File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 452, in initialize_subcommand
    self.subapp.initialize(argv)
  File "<decorator-gen-6>", line 2, in initialize
  File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\jupyter_core\application.py", line 238, in initialize
    self.parse_command_line(argv)
  File "c:\python\python37\lib\site-packages\install\createKernel.py", line 151, in parse_command_line
    self.download_spark(silent, silent_spark_version)
  File "c:\python\python37\lib\site-packages\install\createKernel.py", line 408, in download_spark
    temp_file = self.download_file(spark_download_url)
  File "c:\python\python37\lib\site-packages\install\createKernel.py", line 455, in download_file
    raise Exception("{}".format(response.status_code))
Exception: 404
SPARK_HOME will be set to C:\Users\jerry\pixiedust\bin\spark\spark-2.3.0-bin-hadoop2.7
Downloading Spark 2.3.0


```

**Comment by jerryasher** [comment]
Just for completeness, I retried this on Ubuntu 18.04 using Python 3.7 and that failed.

+ issue in the released version of astunparse re: async is now a keyword not a variable

once that was working, the installation seemed to proceed and work okay downloading everything into my directory, but then when running against the provide pixiedust sample notebooks, it failed when 'sc' would somehow not get defined instead of being a valid ScalaContext (or so I gather.)

**Comment by jaroli** [comment]
@jerryasher I didn't see the `astunparse` error before, but that `sc` is not defined happened to me several times (on Ubuntu 18 server, and Linux Mint). Maybe we should open a new issue for that one?

**Comment by jerryasher** [comment]
That is probably the right thing to do, but I am going to move on for now. I was looking to pixiedust for a jupyter debugger, and what can I say, a jupyter debugger would be great but I seem to be installing a 747 to get it to work.

(the astunparse thing is an issue if python 3.7 is used where `async` is now a keyword and so can't be used as a variable, and the work around for now is to install the latest version of astunparse from its github repo)

tl;dr

+ spark 2.3.0 installation fails with 404
+ spark 2.2.0 windows installation fails with perm errors
+ spark 2.2.0 ubuntu installation fails on python 3.7 with async keyword problem in astunparse
+ spark 2.2.0 ubuntu installation then fails on python 3.7 with scala context issue

I'll try again in six months.

**Comment by jaroli** [comment]
Thanks for the laugh with that 747 comment :)

**Comment by jakubLangr** [comment]
+1 exact same error on MacOS

**Comment by NateZimmer** [comment]
windows + python 3.7
- 2.3.2 fails to download 
- 2.2 downloads, fails for permissions error 

Installed pixiedust 1.1.18 via

```
pip install pixiedust
```
Attempted to install in jupyter
```
jupyter pixiedust install
```

#744 was suppose to fix this in version 1.1.15 so i'm abit confused :/

**Comment by mujina93** [comment]
Same failure here with

centos 7 + python 3.7.3

trying to install spark 2.3.2.

Installing 2.2.0 works fine.

When trying to install 2.3.2 I got:

```
Traceback (most recent call last):
  File "/home/mpumicity/anaconda3/bin/jupyter-pixiedust", line 10, in <module>
    sys.exit(main())
  File "/home/mpumicity/anaconda3/lib/python3.7/site-packages/install/pixiedustapp.py", line 41, in main
    PixiedustJupyterApp.launch_instance()
  File "/home/mpumicity/anaconda3/lib/python3.7/site-packages/traitlets/config/application.py", line 657, in launch_instance
    app.initialize(argv)
  File "</home/mpumicity/anaconda3/lib/python3.7/site-packages/decorator.py:decorator-gen-2>", line 2, in initialize
  File "/home/mpumicity/anaconda3/lib/python3.7/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/home/mpumicity/anaconda3/lib/python3.7/site-packages/traitlets/config/application.py", line 296, in initialize
    self.parse_command_line(argv)
  File "</home/mpumicity/anaconda3/lib/python3.7/site-packages/decorator.py:decorator-gen-4>", line 2, in parse_command_line
  File "/home/mpumicity/anaconda3/lib/python3.7/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/home/mpumicity/anaconda3/lib/python3.7/site-packages/traitlets/config/application.py", line 514, in parse_command_line
    return self.initialize_subcommand(subc, subargv)
  File "</home/mpumicity/anaconda3/lib/python3.7/site-packages/decorator.py:decorator-gen-3>", line 2, in initialize_subcommand
  File "/home/mpumicity/anaconda3/lib/python3.7/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/home/mpumicity/anaconda3/lib/python3.7/site-packages/traitlets/config/application.py", line 452, in initialize_subcommand
    self.subapp.initialize(argv)
  File "</home/mpumicity/anaconda3/lib/python3.7/site-packages/decorator.py:decorator-gen-6>", line 2, in initialize
  File "/home/mpumicity/anaconda3/lib/python3.7/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/home/mpumicity/anaconda3/lib/python3.7/site-packages/jupyter_core/application.py", line 238, in initialize
    self.parse_command_line(argv)
  File "/home/mpumicity/anaconda3/lib/python3.7/site-packages/install/createKernel.py", line 151, in parse_command_line
    self.download_spark(silent, silent_spark_version)
  File "/home/mpumicity/anaconda3/lib/python3.7/site-packages/install/createKernel.py", line 408, in download_spark
    temp_file = self.download_file(spark_download_url)
  File "/home/mpumicity/anaconda3/lib/python3.7/site-packages/install/createKernel.py", line 455, in download_file
    raise Exception("{}".format(response.status_code))
Exception: 404
```

**Comment by alexlopespereira** [comment]
> 
> 
> I am receiving this error as well.
> 
> I am using CPython native for Windows 3.7 on Win 10
> 
> ```
> jerry@Dell5570 MINGW64 ~
> $ pip list | grep jupy
> jupyter            1.0.0
> jupyter-client     5.2.3
> jupyter-console    5.2.0
> jupyter-core       4.4.0
> 
> jerry@Dell5570 MINGW64 ~
> $ python --version
> Python 3.7.0
> 
> jerry@Dell5570 MINGW64 ~
> $ pip list | grep req
> requests           2.19.1
> 
> jerry@Dell5570 MINGW64 ~
> $
> ```
> 
> Specifically, the stacktrace looks like:
> 
> ```
> $ jupyter pixiedust install
> Step 1: PIXIEDUST_HOME: C:\Users\jerry\pixiedust
>         Keep y/n [y]? y
> Step 2: SPARK_HOME: C:\Users\jerry\pixiedust\bin\spark
>         Keep y/n [y]? y
> Directory C:\Users\jerry\pixiedust\bin\spark does not contain a valid SPARK install
>         Download Spark y/n [y]? y
> What version would you like to download? 1.6.3, 2.0.2, 2.1.0, 2.2.0, 2.3.0 [2.3.0]:
> Traceback (most recent call last):
>   File "c:\python\python37\lib\runpy.py", line 193, in _run_module_as_main
>     "__main__", mod_spec)
>   File "c:\python\python37\lib\runpy.py", line 85, in _run_code
>     exec(code, run_globals)
>   File "C:\Python\Python37\Scripts\jupyter-pixiedust.EXE\__main__.py", line 9, in <module>
>   File "c:\python\python37\lib\site-packages\install\pixiedustapp.py", line 41, in main
>     PixiedustJupyterApp.launch_instance()
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 657, in launch_instance
>     app.initialize(argv)
>   File "<decorator-gen-2>", line 2, in initialize
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 87, in catch_config_error
>     return method(app, *args, **kwargs)
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 296, in initialize
>     self.parse_command_line(argv)
>   File "<decorator-gen-4>", line 2, in parse_command_line
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 87, in catch_config_error
>     return method(app, *args, **kwargs)
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 514, in parse_command_line
>     return self.initialize_subcommand(subc, subargv)
>   File "<decorator-gen-3>", line 2, in initialize_subcommand
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 87, in catch_config_error
>     return method(app, *args, **kwargs)
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 452, in initialize_subcommand
>     self.subapp.initialize(argv)
>   File "<decorator-gen-6>", line 2, in initialize
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 87, in catch_config_error
>     return method(app, *args, **kwargs)
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\jupyter_core\application.py", line 238, in initialize
>     self.parse_command_line(argv)
>   File "c:\python\python37\lib\site-packages\install\createKernel.py", line 151, in parse_command_line
>     self.download_spark(silent, silent_spark_version)
>   File "c:\python\python37\lib\site-packages\install\createKernel.py", line 408, in download_spark
>     temp_file = self.download_file(spark_download_url)
>   File "c:\python\python37\lib\site-packages\install\createKernel.py", line 455, in download_file
>     raise Exception("{}".format(response.status_code))
> Exception: 404
> SPARK_HOME will be set to C:\Users\jerry\pixiedust\bin\spark\spark-2.3.0-bin-hadoop2.7
> Downloading Spark 2.3.0
> ```
> 
> Installing Spark 2.2.0 fails with a permissions problem:
> 
> ```
> $ jupyter pixiedust install
> Step 1: PIXIEDUST_HOME: C:\Users\jerry\pixiedust
>         Keep y/n [y]? y
> Step 2: SPARK_HOME: C:\Users\jerry\pixiedust\bin\spark
>         Keep y/n [y]? y
> Directory C:\Users\jerry\pixiedust\bin\spark does not contain a valid SPARK install
>         Download Spark y/n [y]? y
> What version would you like to download? 1.6.3, 2.0.2, 2.1.0, 2.2.0, 2.3.0 [2.3.0]: 2.2.0
> SPARK_HOME will be set to C:\Users\jerry\pixiedust\bin\spark\spark-2.2.0-bin-hadoop2.7
> Downloading Spark 2.2.0
> Traceback (most recent call last):
>   File "c:\python\python37\lib\runpy.py", line 193, in _run_module_as_main
>     "__main__", mod_spec)
>   File "c:\python\python37\lib\runpy.py", line 85, in _run_code
>     exec(code, run_globals)
>   File "C:\Python\Python37\Scripts\jupyter-pixiedust.EXE\__main__.py", line 9, in <module>
>   File "c:\python\python37\lib\site-packages\install\pixiedustapp.py", line 41, in main
>     PixiedustJupyterApp.launch_instance()
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 657, in launch_instance
>     app.initialize(argv)
>   File "<decorator-gen-2>", line 2, in initialize
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 87, in catch_config_error
>     return method(app, *args, **kwargs)
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 296, in initialize
>     self.parse_command_line(argv)
>   File "<decorator-gen-4>", line 2, in parse_command_line
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 87, in catch_config_error
>     return method(app, *args, **kwargs)
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 514, in parse_command_line
>     return self.initialize_subcommand(subc, subargv)
>   File "<decorator-gen-3>", line 2, in initialize_subcommand
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 87, in catch_config_error
>     return method(app, *args, **kwargs)
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 452, in initialize_subcommand
>     self.subapp.initialize(argv)
>   File "<decorator-gen-6>", line 2, in initialize
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\traitlets\config\application.py", line 87, in catch_config_error
>     return method(app, *args, **kwargs)
>   File "C:\Users\jerry\AppData\Roaming\Python\Python37\site-packages\jupyter_core\application.py", line 238, in initialize
>     self.parse_command_line(argv)
>   File "c:\python\python37\lib\site-packages\install\createKernel.py", line 151, in parse_command_line
>     self.download_spark(silent, silent_spark_version)
>   File "c:\python\python37\lib\site-packages\install\createKernel.py", line 410, in download_spark
>     self.extract_temp_file(temp_file, self.spark_home)
>   File "c:\python\python37\lib\site-packages\install\createKernel.py", line 479, in extract_temp_file
>     tar = tarfile.open(temp_file.name, "r:gz")
>   File "c:\python\python37\lib\tarfile.py", line 1589, in open
>     return func(name, filemode, fileobj, **kwargs)
>   File "c:\python\python37\lib\tarfile.py", line 1636, in gzopen
>     fileobj = gzip.GzipFile(name, mode + "b", compresslevel, fileobj)
>   File "c:\python\python37\lib\gzip.py", line 163, in __init__
>     fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')
> PermissionError: [Errno 13] Permission denied: 'C:\\Users\\jerry\\AppData\\Local\\Temp\\tmpbr_7nwlb.tgz'
> Extracting Spark 2.2.0 to C:\Users\jerry\pixiedust\bin\spark
> ```
> 
> Rerunning "as Administrator" didn't help. Neither did switching from a git bash for windows shell to a powershell shell.

It happens because in Windows 10 the directory C:\Users\alex\AppData\Local\Temp is read only.
It might be defined in some environmental variable. Is it correct? If we change the destination directory for a folder with proper permissions we could install it.
